### ISSUE TITLE
Study alternative containers for Storage

### ISSUE THREAD

**Issue by damip** [massalabs/massa#2344]
The shared Storage has few concurrent writes by a couple of threads, heavy concurrent reads by many threads.

We are currently planning on using `Arc<RwLock<PreHashMap<BlockId, Arc<RwLock<BlockInfo>>>>` as the underlying storage container.

Now I recently watched this: https://www.youtube.com/watch?v=HJ-719EGIts

Basically it could be possible to do most of the work we want to do lock-free, with much higher performance !

Then I looked for stuff and found crates proposing this kind of stuff:
* https://docs.rs/flurry/latest/flurry/
* https://github.com/xacrimon/dashmap  (with comparative benchmark:  https://github.com/xacrimon/conc-map-bench  )
* https://github.com/jonhoo/left-right
* https://docs.rs/evmap/latest/evmap/

It looks like those structures can yield an order-of-magnitude performance improvement under the right conditions (that look like the conditions we have).

However, depending on the implementation, there could be some unexpected problems such as a temporary lack of read/write consistency.

We need to:
* study the proposed containers (and look for others)
* get a more precise idea of the Read/Write concurrency and throughput expectations in our architecture. Maybe wait to have the first version of Storage implemented to do an actual benchmark
* try changing the underlying Storage container with the various options and bench the impact on performance

**Comment by gterzian** [comment]
> Basically it could be possible to do most of the work we want to do lock-free, with much higher performance !

I am doubtful of the higher performance of these things. 

For example, see the below excerpt from https://www.chromium.org/developers/lock-and-condition-variable/

> Programmers assume that locking is expensive, and that using atomic
operations will be more efficient. But in reality, acquiring and releasing a
lock is cheaper than a cache miss; attention to cache behaviour is usually a
more fruitful way to improve performance. Furthermore, lock-free data
structures are often more expensive than using locks. This is because a lock
allows arbitrary changes to be made to a complex data structure; if the same
changes must be made without a lock, the result is likely to take more
atomic read-modify-write instructions, not fewer.

In our specific situation, we have the below structure:

`Arc<RwLock<Map<BlockId, Arc<RwLock<StoredBlock>>>>>`

Most of the "work" will occurs while locking an individual `StoredBlock`, so there should be little contention on the lock around the top-level map. 

That lock around the top-level map could be replaced by using some sort of lock-free map, however we should consider:

1. Probably losing the optimization of our current pre-hashed maps.
2. That going lock-free there probably doesn't really give any relevant performance boost.
3. That we also may want the synchronization that comes with a lock. For example if a block is pruned, we may not want it to be able to be read logically after that point. Lock-free structures may have looser synchronization, and/or if they do allow such "hard" synchronization, they might not be any faster than simply a lock around a map.

**Comment by AurelienFT** [comment]
I don't have a full vision of the code so I can't talk about if the synchronization will be a real problem.

But regarding this SO answer : https://stackoverflow.com/a/5680988 it may depend on the algorithm and also in the cases. If we look at this benchmark : https://github.com/xacrimon/conc-map-bench with a lot of reads and less writing it seems that we can have a huge improvement.

I think we are reading a lot more than writing so maybe it will be worth but as it's said in the SO answer : 

> you really don't know until you measure.

So I saw that you (@gterzian) already have some branches with implementations of a storage using `RwLock` can you tell me which one is the most update and the most relevant to test on (if there is one) and I will try to add a "relevant" benchmark around it and try different storages.

**Comment by AurelienFT** [comment]
I tried this branch `shared_storage_protocol` and follow this doc to run a dummy network : https://github.com/massalabs/massa/wiki/Dummy-network-generation

But I'm facing a lot of call to `retrieve_block` that are `unwrap` returning None and then make the node crash. I tried to fix some but there is some I don't know how to. Do you have an idea on how I can have a node using the shared network (at least in some case) that can run ? (if possible)

**Comment by AurelienFT** [comment]
Tried to run the benchmark that is pretty cool : https://github.com/ibraheemdev/conc-map-bench (this fork have a fix for flurry) with sha256 with several crates (bitcoin_hashes and RustCrypto) but they don't implement `Hasher` trait.

Following this issue : https://github.com/RustCrypto/hashes/issues/49 it's not a good way to implement this trait for cryptographic hashes. Also, following the SO answer above (and my mind) the best way is to test in real condition so I will create a benchmark on the existing code for storage (when I will be able to run it (cf. comment just above))

**Comment by AurelienFT** [comment]
Btw @damip left-right is a child project of evmap : https://github.com/jonhoo/left-right/commit/abe3769feda942d26b6f24eac99035e1c80e04e0

**Comment by AurelienFT** [comment]
I have fork and run the benchmark with the hash of Massa and prehash on `massa_models`. Here is the results:

## Scenario Heavy Read : 
```
read   98%
insert  1%
remove  1%
```
![ReadHeavy latency](https://user-images.githubusercontent.com/32803821/156220265-f692d75a-b553-4af1-b489-8164178bf8a8.svg)
![ReadHeavy throughput](https://user-images.githubusercontent.com/32803821/156220328-477a422e-b1ac-4569-9976-45686c8a43e6.svg)

## Scenario Exchange : 
```
read    10%
insert  40%
remove  40%
update  10%
```
![Exchange latency](https://user-images.githubusercontent.com/32803821/156220596-00628b5d-a11e-4633-8fe9-19bfe4a4e09a.svg)
![Exchange throughput](https://user-images.githubusercontent.com/32803821/156220600-0a630652-5a36-44cc-84b6-fdf3d853ec1f.svg)


## Scenario rapid grow : 
```
read    5%
insert 80%
remove  5%
update 10%
```
![RapidGrow latency](https://user-images.githubusercontent.com/32803821/156220754-09a85676-3da2-4311-9e59-94d2c0cfca30.svg)
![RapidGrow throughput](https://user-images.githubusercontent.com/32803821/156220756-4df9c0fc-0a96-4386-8921-f935416c0c89.svg)

I am working on a fork of bustle (crate used in the benchmark) to split the read and write operations in specific threads and be able to specify how many read/write threads we want.

Links : 
- Fork of the benchmark : https://github.com/AurelienFT/conc-map-bench
- Fork of bustle : https://github.com/AurelienFT/bustle

**Comment by AurelienFT** [comment]
I have successfully modified the two crates to allow a number of read and write threads. It's a bit "hacky" and if we want to contribute it upstream I will have to refactor it a bit but I don't think it's our priority right now.

Here is the results : (to read the X axis: 2 first digits are number of read threads, 0 separate, 2 last digits numbers of write threads)

## Scenario Heavy Read : 
```
read   98%
insert  1%
remove  1%
```
![ReadHeavy throughput](https://user-images.githubusercontent.com/32803821/156376752-86adb146-cbec-44f3-9fb6-5dfde1e34abc.svg)
![ReadHeavy latency](https://user-images.githubusercontent.com/32803821/156376749-6b667d1a-7676-495c-8ff4-07528e155f25.svg)


## Scenario Exchange : 
```
read    10%
insert  40%
remove  40%
update  10%
```
![Exchange throughput](https://user-images.githubusercontent.com/32803821/156406455-2ba92120-a24d-4cd6-b90e-e423f67aa719.svg)
![Exchange latency](https://user-images.githubusercontent.com/32803821/156406450-7ff9b88f-b98b-491f-9024-5705bd5f314b.svg)

## Scenario rapid grow : 
```
read    5%
insert 80%
remove  5%
update 10%
```
(COMING BENCHMARK RUNNING CURRENTLY)

If you want me to rerun with different data/number of threads it's possible you can also run it directly with my own fork of the crates : 
- Fork of the benchmark : https://github.com/AurelienFT/conc-map-bench
- Fork of bustle : https://github.com/AurelienFT/bustle

To run the benchmark on the fork you can run this command : 
```
./scripts/bench.bash --read-threads="9" --write-threads="2" --read-threads="8"  --write-threads="2" --read-threads="7"  --write-threads="2" --read-threads="7"  --write-threads="3" --read-threads="7"  --write-threads="4"
```

The a --read-threads  flag must always be paired with a --write-threads which will create a pair of test for benchmark and as shown above you can add multiples. (Will try to change to accept comma list directly).

To render the plots : 
```
./scripts/plot.bash
```

**Comment by adrien-zinger** [comment]
Dashmap works with sharding, that's smart, not magic and scalable!! :bearded_person: :love_you_gesture:

**Comment by AurelienFT** [comment]
> Dashmap works with sharding, that's smart, not magic and scalable!! 🧔 🤟

Flurry also :

> This map usually acts as a binned (bucketed) hash table. Each key-value mapping is held in a BinEntry

**Comment by AurelienFT** [comment]
### New benchamrk with random data as value instead of 0

Here is the results : 

## Scenario Heavy Read : 
```
read   98%
insert  1%
remove  1%
```
![ReadHeavy throughput](https://user-images.githubusercontent.com/32803821/156558831-934c24b9-c6fd-4ea8-bd00-2c48cbaf213a.svg)
![ReadHeavy latency](https://user-images.githubusercontent.com/32803821/156558829-baf77f8c-4d32-4471-b67f-cde50bcefbdf.svg)

## Scenario Exchange : 
```
read    10%
insert  40%
remove  40%
update  10%
```
![Exchange throughput](https://user-images.githubusercontent.com/32803821/156558881-076d1124-990c-422e-9dd9-2e5e4e9216b5.svg)
![Exchange latency](https://user-images.githubusercontent.com/32803821/156558875-4fb34abc-a9b2-4c0d-ab9f-a2aecb007b92.svg)

## Scenario rapid grow : 
```
read    5%
insert 80%
remove  5%
update 10%
```
![RapidGrow throughput](https://user-images.githubusercontent.com/32803821/156558951-6d0e1d2e-721f-46e9-9e7f-2055ad38774d.svg)
![RapidGrow latency](https://user-images.githubusercontent.com/32803821/156558943-b1db332a-4431-4a90-93d9-536354259ec7.svg)

**Comment by gterzian** [comment]
@AurelienFT Thank you for your enthousiasm. 

I would like to reiterate what I wrote above:

> Could be worth running benchmarks after the refactoring has been completed, and after having determined that the application spends a significant amount of time in that read/write workflow(either manipulating the map, or waiting on the lock).

For clarity, this means that I don't think we should be making any efforts on this until the following has been reached:

1. We have (largely) finished the integration of storage into the project.
2. We have noticed that the application spends a significant amount of time waiting on the lock on the top-level map becoming available. 

Point 2 is especially important, since we can always optimize things, but most of it is wasted effort since if it will not materially affect overall performance of the application. 

> if you have some tasks to help you on the integration of the storage so that I can help you

Appreciated, the main discussion would be over at https://github.com/massalabs/massa/issues/2178
I think a good start would be to familiarize yourself with the various issues and currently open PR. I think there will be more specific tasks next week when some initial work is finished. 

> because the more usage we have the more the bench will be

As per above, the goal is not to optimize this, unless we notice it becomes a problem :)

As an example, we can have a 10 times improvement on "something", however this is meaningless if the app spends only 0,00001% of its time doing that "thing".
